keras_cv/layers/vit_layers.py

```diff
@@ -0,0 +1,204 @@
+        - https://github.com/keras-team/keras-io/blob/master/examples/vision/image_classification_with_vision_transformer.py
+
+    The layer expects a batch of input images and returns batches of patches.
+
+    args:
+        - patch_size: the size (patch_size, patch_size) of each patch created from the image
+
+    Basic usage:
+
+    ```
+    img = url_to_array(url)
+    batch_img = np.expand_dims(img, 0)
+
+    patch_size = 16
+    patches = keras_cv.layers.Patching(patch_size)(batch_img)
+    print(f"Image size: {batch_img.shape}")                # Image size: (1, 224, 224, 3)
+    print(f"Patch size: {(patch_size, patch_size)}") # Patch size: (16, 16)
+    print(f"Patches: {patches.shape[1]}")            # Patches: 196
+    ```
+    """
+
+    def __init__(self, patch_size, **kwargs):
+        super().__init__(**kwargs)
+        self.patch_size = patch_size
+
+    def call(self, images):
+        batch_size = tf.shape(images)[0]
+        patches = tf.image.extract_patches(
+            images=images,
+            sizes=[1, self.patch_size, self.patch_size, 1],
+            strides=[1, self.patch_size, self.patch_size, 1],
+            rates=[1, 1, 1, 1],
+            padding="VALID",
+        )
+        patch_dims = patches.shape[-1]
+        patches = tf.reshape(
+            patches, [batch_size, patches.shape[-2] * patches.shape[-2], patch_dims]
+        )
+        return patches
+
+    def get_config(self):
+        config = {
+            "patch_size": self.patch_size,
+        }
+        base_config = super().get_config()
+        return dict(list(base_config.items()) + list(config.items()))
+
+
+@tf.keras.utils.register_keras_serializable(package="keras_cv")
+class PatchEmbedding(layers.Layer):
+    """
+    Layer to concat a class token, positionally embed and create a projection of patches made with the `Patching` layer
+    for Vision Transformers from:
+        - An Image is Worth 16x16 Words: Transformers for Image Recognition at Scale
+        by Alexey Dosovitskiy et al. (https://arxiv.org/abs/2010.11929)
+
+    Based on Khalid Salama's implementation for:
+        - https://github.com/keras-team/keras-io/blob/master/examples/vision/image_classification_with_vision_transformer.py
+
+    args:
+        - project_dim: the dimensionality of the project_dim
+
+    Basic usage:
+
+    ```
+    patches = keras_cv.layers.Patching(patch_size)(batch_img)
+
+    project_dim = 1024
+    num_patches = patches.shape[1] # 196
+
+    encoded_patches = keras_cv.layers.PatchEmbedding(project_dim=project_dim)(patches)
+    print(encoded_patches.shape) # (1, 197, 1024)
+    ```
+    """
+
+    def __init__(self, project_dim, **kwargs):
+        super().__init__(**kwargs)
+        self.project_dim = project_dim
+        self.linear_projection = layers.Dense(self.project_dim)
+
+    def build(self, input_shape):
+        self.class_token = tf.random.normal([1, 1, input_shape[-1]], name="class_token")
+        self.num_patches = input_shape[1]
```

Comment:
isn't the number of patches dynamic?

Reply:
The number is inherited from the input, the class token is pre-pended, and if the interpolation flag is set - the number of patches is changed, and the positional embeddings are interpolated for the new shape. I'd say it's dynamic

Reply:
right, as you mentioned, it depends on the input (or the output of the Patching layer). Given the patch size is fixed, this means the number of patches changes as the H / W changes

Reply:
Hmm, yes - this could pose an issue with datasets where the sizes change between batches. I don't know how common these are, but we'd ideally want to allow for those as well. Good catch!
I'm not sure how we'd make it fully dynamic (None types tend to cause build-type EagerTensor exceptions) but I'll look into how we can make it robust to datasets with batches of images of different shapes

Reply:
That depends on the training script, some of our existing models require fixed shape, so they're resized and cropped. You might need to read into the the paper to see if they are doing something similar, or actually passing different shapes across batches -- whatever gets you best results.
But in general, supporting dynamic input shape is desired because users can structure their own dataset to do so

Reply:
+1, supporting dynamic input shape when possible is optimal.